### ISSUE TITLE
push notification queue

### ISSUE THREAD

**Issue by fairsayan** [argon/node-apn#281]
I'm testing apn module sending multiple notifications together.
I tried with multiple connections and with the same connection and the behaviour is the same:
when I send many notifications (8 - 10 of them) in a short period they are sent all together, after that if I try to send again the same amount of notifications they remain all pending 
apn 1 left to send +0ms
apn Shutdown pending +1ms
(this is with multiple connections otherwise I would read "apn 5 left to send" depending on the amount of pending notifications of that connection)

but if I restart node and I try to send them are sent immediately again. I didn't check the code yet, but the behaviour looks like a problem with the buffer system that is not drained properly. So I tried to disable the buffersNotifications option but nothing change. 
Any clue about it?

**Comment by argon** [comment]
It looks like you have called `connection.shutdown()` somewhere. This closes the module down completely so it will never send any more pushes. It sounds like you don't want to call that function if you want the process to be long-running

―A

**Comment by fairsayan** [comment]
I tried both way with and without shutdown, same effect. The only difference is that without it the process never exit and every new connection is a deadlock.
BTW why a shutdown on a connection should affect a new one? I mean if a create a new connection for every sendPushNotification then even if there is a shutdown it doesn't stop following new connections, right?

**Comment by argon** [comment]
why are you creating new connections every time? That is likely to get you blocked by Apple. You should create one connection and send all your notifications over it. If you set the `maxSockets` option then node-apn will create multiple underlying sockets and ensure that they are used efficiently.

However, without further information I can't help with the blocking problem. If you can post an example of the code you're using I may be able to help further.

**Comment by fairsayan** [comment]
ok, when a little bit deeper in our code. The problem is affecting the tls.connect in your socket.js that never calls the callback. This happen only if I run the code from mocha test (I'm using grunt as task manager to run tests every time I modify the files) and it is independent if it is the first time the code is run or not. I verified that for both successful cases and failures socketOptions are the same.
I tried to reproduce the same behaviour in a simplified version to send you, but I wasn't able to do it.
I close this issue.
